### PR TITLE
feat: Implement strongly typed function signatures for dotnet

### PR DESCRIPTION
**Pull request by garethgeorge**
Implements a strongly typed function signature that provides automatic request deserialization and response serialization for dotnet.

### CODE REVIEW

**Comment by jskeet** [comment]
Starting to review and fix nits locally now - but while I remember, we generally find it better to work on forks and create pull requests from there, rather than accumulating branches on the "main" fork here. Not a problem, but if you could create a fork to work on for the next PR, that'd be great :)

File: src/Google.Cloud.Functions.Hosting/HostingInternals.cs
Comment (jskeet) [inline src/Google.Cloud.Functions.Hosting/HostingInternals.cs]: I've added a comment here to indicate that users can specify their own readers/writers, and added a new test for that.

**Review by jskeet — COMMENTED** [review]
I've added two extra commits with a bunch of changes - please review them carefully.
The first contains "real" changes (mostly naming, admittedly). The second is *just* changing to use file-scoped namespaces for the new code, but if I included that with the first commit, it would make it much harder to read.

If you're happy with the changes, we can then add another commit in the examples directory, then squash and merge.

File: src/Google.Cloud.Functions.Framework/TypedFunctionAdapter.cs
Comment (garethgeorge) [reply src/Google.Cloud.Functions.Framework/TypedFunctionAdapter.cs]: Right, I hadn't thought much about the fact that theres now no way to indicate a legitimate 500 error on the request parsing edge. For the initial implementation, it is my hope that *most* exceptions thrown in formatters will be legitimate parse errors or the such so we throw the 400 as a reasonable best guess.

@anniefu has also mentioned asks for exceptions that give some control over the HTTP status returned, it's something that would be neat to look at eventually.

File: src/Google.Cloud.Functions.Framework/IHttpRequestReader.cs
Comment (garethgeorge) [reply src/Google.Cloud.Functions.Framework/IHttpRequestReader.cs]: It's an interesting thought, CancellationTokens are used to abort an operation if it's long running and the HTTP request dies? I'd say I'm also ambivalent. 

The cases I can imagine where it might be useful would be borderline applications of formatters i.e. augmenting the request with data fetched from a Cloud Storage bucket (or some similarly slow and expensive network operation you'd want to be able to cancel). In the typical case request decoding / response encoding should be so fast I can't imagine too much value.

WDYT?

**Comment by garethgeorge** [comment]
Thanks Jon! Did a pass over your revisions, they LGTM. Appreciate the places you've revised the implementation and test coverage. 

Also good catch w.r.t. throwing an error on a null request payload in the default json reader, it's a good idea. That's something I'll aim to make sure we're consistent about (where we can be) across the other languages we're adding this feature to.

**Comment by jskeet** [comment]
> Also good catch w.r.t. throwing an error on a null request payload in the default json reader, it's a good idea. 

Your code was already doing that, actually - I just added a test for it, and made the exception a bit clearer.

File: src/Google.Cloud.Functions.Framework/IHttpRequestReader.cs
Comment (jskeet) [reply src/Google.Cloud.Functions.Framework/IHttpRequestReader.cs]: I agree that usually it'll be fast, in terms of processing. If a large result is generated, then it would be good to abort that early, but I'd expect that if the request is aborted, that output stream will become invalid anyway.

Let's go without for now... introducing it later would be a breaking change, but we can always introduce other interfaces to implement, and work things out that way.

**Review by jskeet — COMMENTED** [review]
Given that a fair amount of the code is now mine, I'll ask for a second opinion to actually approve before we squash and merge.

File: src/Google.Cloud.Functions.Framework/ITypedFunction.cs
Comment (amanda-tarafa) [inline src/Google.Cloud.Functions.Framework/ITypedFunction.cs:1]: nit: 2023, and maybe check elsewhere

File: src/Google.Cloud.Functions.Framework/ITypedFunction.cs
Comment (amanda-tarafa) [inline src/Google.Cloud.Functions.Framework/ITypedFunction.cs]: nit: empty docs

File: src/Google.Cloud.Functions.Framework/ITypedFunction.cs
Comment (amanda-tarafa) [inline src/Google.Cloud.Functions.Framework/ITypedFunction.cs]: Do we really need these constraints?

File: src/Google.Cloud.Functions.Hosting/HostingInternals.cs
Comment (amanda-tarafa) [inline src/Google.Cloud.Functions.Hosting/HostingInternals.cs:185]: thought: This seems a little random? Maybe consider adding a utility for generating the System.Text.Json reader/writer and find a way to annotate the typed function to specify those. It would be similar to the formatter for Cloud Events.

File: src/Google.Cloud.Functions.Framework/ITypedFunction.cs
Comment (amanda-tarafa) [reply src/Google.Cloud.Functions.Framework/ITypedFunction.cs]: Oh never mind, I've seen these have been removed in later commits.

File: src/Google.Cloud.Functions.Framework/IHttpRequestReader.cs
Comment (amanda-tarafa) [reply src/Google.Cloud.Functions.Framework/IHttpRequestReader.cs]: There's no need to explicitly pass the cancellation token. HttpRequest and HttpResponse both have access to the HttpContext, which in turn has the `CancellationToken HttpContext.RequestAborted` property.

The alternative I would consider for the reader and the writer is to actually pass the HttpContext directly. That makes it clearer that the whole thing is available for using in serialization/deserialization. There are other elements of HttpContext that may be useful for deserialization, like the Items collection.

File: src/Google.Cloud.Functions.Framework/ITypedFunction.cs
Comment (jskeet) [reply src/Google.Cloud.Functions.Framework/ITypedFunction.cs:1]: Thanks, will fix - it looks like we're missing headers in another couple of files too.

File: src/Google.Cloud.Functions.Hosting/HostingInternals.cs
Comment (jskeet) [reply src/Google.Cloud.Functions.Hosting/HostingInternals.cs:185]: Unlike the CloudEvents scenario, we don't have any common non-FF assembly that could contain an attribute specifying what reader/writer should be used.

But we're expecting JSON to be the default serialization format for most things. We may provide docs with an example of Json.NET, as that'll be reasonably common too.

File: src/Google.Cloud.Functions.Framework/IHttpRequestReader.cs
Comment (jskeet) [reply src/Google.Cloud.Functions.Framework/IHttpRequestReader.cs]: Ah, hooray. Good to know we don't need to pass anything else.

But I prefer keeping the request/response directly - it emphasizes what the code is expected to work with. (It would be relatively unusual for a request reader to need to reach into the context, for example.) So long as they *can* do it, I think that's fine.

**Review by jskeet — APPROVED** [review]
All copyright files checked now. I've squashed it into a single commit - will rebase and merge shortly. Expect to see buildpack integration tests failing - that's due to our SDK being a bit later than the one in buildpacks at the moment.

**Pull request merged**
